lib/nvml.ml

```diff
@@ -136,13 +138,59 @@ let get_vgpu_for_uuid iface vgpu_uuid vgpus =
     )
     vgpus
 
-(* mock implementation *)
-module NVML = struct
-  let attach () = ()
+module NVML : sig
+  val attach : unit -> unit
+
+  val detach : unit -> unit
+
+  val is_attached : unit -> bool
+
+  val get : unit -> interface option
+end = struct
+  let interface = ref (None : interface option)
+
+  let mx = Mutex.create ()
 
-  let detach () = ()
+  let finally = Xapi_stdext_pervasives.Pervasiveext.finally
 
-  let is_attached () = true
+  let with_mutex mx f =
+    Mutex.lock mx ;
+    finally f (fun () -> Mutex.unlock mx)
+
+  let open_nvml_interface () =
+    let interface = library_open () in
+    try init interface ; interface with e -> library_close interface ; raise e
+
+  let get () = !interface
+
+  let attach () =
+    with_mutex mx @@ fun () ->
+    match !interface with
+    | Some _ ->
+        D.warn "Nvml library attach: library already attached"
```

Comment:
Attach and detach are idempotent, which avoids trouble with interference between the RRD thread and the RCP calls.